### ISSUE TITLE
Difficulties replicating pre-trained results from scratch

### ISSUE THREAD

**Issue by psturmfels** [songlab-cal/tape#77]
I've been trying to replicate or at least get close to the results you highlight in the paper, with the understanding that there may be some differences from the TensorFlow version.

However, I've been unable to train the transformer with the masked language modeling task successfully - perplexity seems to initially lower, but as the learning rate ramps up perplexity jumps up and never really seems to go back down. I'm using the following settings:

```
{
  "base_model": "transformer",
  "attention_probs_dropout_prob": 0.1,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "max_position_embeddings": 8192,
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "type_vocab_size": 1,
  "vocab_size": 30,
  "layer_norm_eps": 1e-12
}
```

And running the code as follows:
```
python run_distributed.py
transformer masked_language_modeling
--gradient_accumulation_steps 64
--batch_size 512
--learning_rate 1e-3
--nproc_per_node 4
--fp16
```

with the rest of the parameters set as the default ones. 

However, my training curves start to look like this: where the training perplexity initially lowers but then jumps up during the learning rate warm up phase. Just to be sure, I've tried it with three separate random initializations and seen the same pattern.
![Screen Shot 2020-07-14 at 2 48 07 PM](https://user-images.githubusercontent.com/10600588/87479841-09be4780-c5e1-11ea-860b-73af2ad9e3e8.png)

Do you have any advice for parameter settings I might be setting incorrectly? And what did your training curves look like? Is it just a matter of letting it train for a lot longer? For reference, I'm running with 4 Nvidia V100 GPUs. 

Thank you for the great repository by the way - it's been very easy and clear to work with and extend.

**Comment by rmrao** [comment]
Thanks for the interest in the project! I have noticed that the pytorch code is a little more finicky than the tensorflow code (plausibly due to slight differences in the way BERT is implemented). I've had good success with lowering the learning rate to between 1e-4 and 5e-4.

**Comment by psturmfels** [comment]
Thank you for the rapid response! Trying a smaller learning rate certainly makes sense given my training curves - I'll give it a try and report results back here on this issue.

**Comment by psturmfels** [comment]
For the record: setting the maximum learning rate to be 2.5e-4 appears to be on track to closely approximate results from the original paper, whereas for me, 5e-4 was still too large. Thanks again for the help with this issue!
